src/EventProducerService.cpp

```diff
@@ -38,37 +38,65 @@ void EventProducerService::setProducedEvents()
     data[1] = lowByte(module_config->nodeNum);
     data[2] = 0;
     data[3] = i;
- 
-    byte index = i - 1;
+    
+    byte index = module_config->findExistingEventByEv(1, i);
+    if (index >= module_config->EE_MAX_EVENTS)  //event does not exist so creat default
+    {
+      index = module_config->findEventSpace();
+    }
+        
     module_config->writeEvent(index, data);
+    module_config->writeEventEV(index, 1, i);
     module_config->updateEvHashEntry(index);    
   }    
 }
 
+void EventProducerService::recreateDefaultEvent(byte evValue)
+{
+  byte data[4];
+  data[0] = highByte(module_config->nodeNum);
+  data[1] = lowByte(module_config->nodeNum);
+  data[2] = 0;
+  data[3] = evValue;
+  
+  byte index = module_config->findEventSpace();
+  
+  module_config->writeEvent(index, data);
+  module_config->writeEventEV(index, 1, evValue);
+  module_config->updateEvHashEntry(index);
+}
+
+
 void EventProducerService::process(UserInterface::RequestedAction requestedAction)
 {
   // Do this if mode changes from uninitialised to normal
-  if (((uninit) && (module_config->currentMode == MODE_NORMAL)) || module_config->isResetFlagSet())
+  if (((uninit) && (module_config->currentMode == MODE_NORMAL)))
   {
     setProducedEvents();
     uninit = false;
   }
 }
 
-void EventProducerService::sendEvent(bool state, byte index)
+void EventProducerService::sendEvent(bool state, byte evValue)
 {
+  byte index = module_config->findExistingEventByEv(1, evValue);
   if (index < module_config->EE_MAX_EVENTS)
   {
     byte opCode;
     byte nn_en[4];
     module_config->readEvent(index, nn_en);
-    //DEBUG_SERIAL << F("eps>index = ") << index << F(" , Node Number = 0x") << _HEX(nn) << endl;
+    //DEBUG_SERIAL << F("eps>index = ") << index << F(" , Node Number = 0x") << _HEX(nn_en[0]) << _HEX(nn_en[1]) << endl;
     if ((nn_en[0] == 0) && (nn_en[1] == 0))
     {
       opCode = (state ? OPC_ASON : OPC_ASOF);
       nn_en[0] = highByte(module_config->nodeNum);
       nn_en[1] = lowByte(module_config->nodeNum); 
     }
+    else if ((nn_en[0] == 0xff) && (nn_en[1] == 0xff))
+    {
+      recreateDefaultEvent(evValue);
```

Comment:
This looks strange to me. I find it strange to recreate devault events while sending events. 
As I read the code, if the event we are about to send is NN/EN == FFFF (Event is found but not set up(?)) then we recreate the default events. Is this some way to catch a bad configuration? 
Under what conditions whould you enter into this code here? How can we have an event in the table but with unset NN/EN?

Reply:
If you delete a default event, the nn and en values in the events table get set to 0xFF.  However, EVs are not deleted and so if that index position hasn't been overwritten by another teach event the pressing of a switch to re-instate will find this index location with a nn_en of 0xFFFF.  Thinking about it, if that index position has now been re-used, the findExistingEventbyEv() will return an index > EE_MAX_EVENTS and so I need an else statement to call recreateDefaultEvent at the end of the first if statement as well.

Reply:
Yes, I was thinking that the entry in evhashtbl would be cleared which means it wouldn't be found.
Anyway, I think sendEvent() is the wrong place to recreate default events. If you have deleted such an event then you don't want it recreated. I agree that in that case you shouldn't call sendEvent with that EV1 value. Doing so should be considered an error, not recreating the events.
We need another mechanism for recreating the default events. Keep in mind that Ian wants the NNCLR as a way to get rid of all default events. ;) For now NNRSM seems to be the only valid way to get default events back.
Happy to discuss this further.

Reply:
If the NV for a switch (in the case of 4in4out) has a valid setting i.e. 1-4, then pressing the switch will re-instate the default.  This is done using sendEvent as the simplest way.  If the NV is set to 0 or an undefined number (5 upwards in the case of 4in4out) then sendEvent is not sent.  The problem with using the NNRSM route is that you delete everything else that may have been taught!  NNCLR gets rid of all events, not just defaults.  If the NVs are set to 0, then they will not be re-instated.

Reply:
Reading code better now and see that you are only reinstating _one_ event. 
And your comment on the NV controlling the existence of the produced event makes sense.

Reply:
It's good to be challenged on these things as it makes me think about what I've done.  It also helped me pick up an omission that would have become a head-scratching issue in the future.